src/mlpack/core/optimizers/katyusha/katyusha_impl.hpp

```diff
@@ -0,0 +1,188 @@
+/**
+ * @file katyusha_impl.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of Katyusha a direct, primal-only stochastic gradient method.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_KATYUSHA_KATYUSHA_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_KATYUSHA_KATYUSHA_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "katyusha.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+Katyusha::Katyusha(
+    const double stepSize,
+    const double lambda,
+    const double tau,
+    const size_t batchSize,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle) :
+    stepSize(stepSize),
+    lambda(lambda),
+    tau(tau),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double Katyusha::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+  const double tau1 = 0.5 - tau;
+  const double tau2 = 1 + lambda * stepSize;
+
+  // To keep track of where we are and how things are going.
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; i += batchSize)
+  {
+    const size_t effectiveBatchSize = std::min(batchSize, numFunctions - i);
+    overallObjective += function.Evaluate(iterate, i, effectiveBatchSize);
+  }
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat fullGradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  arma::mat y = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat z = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat w = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
```

Comment:
My intention was to use:
```c++
effectiveBatchSize = std::min(batchSize, numFunctions - currentFunction);
overallObjective += function.Evaluate(iterate, currentFunction, effectiveBatchSize);
```
the evaluation after the inner loop as a proxy for the overall performance, we could remove the if condition and evaluate the complete function, but I don't think that's necessary here. Let me know what you think.

> And should we set actualMaxIterations to the user-supplied maxIterations, or to maxIterations / numFunctions?

Yes, I will do that.

Reply:
Okay, I've spent some more time with the code and I think I understand the transformation from `Katyusha1()` a bit better now.

Based on my understanding of `Katyusha1()` (just to be clear I mean Algorithm 3 of https://arxiv.org/pdf/1603.05953.pdf, in case there are different versions of the paper out there), there is an outer loop from line 11 to 23 and an inner loop from line 13 to 21.  The outer loop is performed S times, and updates the full gradient, then updates the parameter with the results of the inner loop (the parameter update can be moved into the inner loop and it is in your implementation).  So I would have expected code kind of like the form...

```
size_t maxOuterIterations = actualMaxIterations / numFunctions; // or something like this
for (size_t s = 0; s < maxOuterIterations; ++s)
{
  // calculate full gradient, print objective or an estimate of it

  // do inner loop stuff
}
```

so what was confusing me was that the outer iteration counter (`i`) is being incremented by `effectiveBatchSize` each time where I was expecting that either `i` would be incremented by `numFunctions` (with a maximum of `actualMaxIterations`), or `i` would be incremented by 1 (with a maximum of `numFunctions / actualMaxIterations` or something like that).  I hope that helps explain my confusion a little bit.

So it seems like the reason you are incrementing by `effectiveBatchSize` and not `numFunctions` is to avoid a potentially costly objective function evaluation every outer iteration, and then the objective will be printed to the user less often (once every `numFunctions / batchSize` outer iterations).  I wonder if we should instead print an objective function estimate based on the single-batch calculation you are doing, or something else.  It is really useful for users to see some kind of objective function so they know the optimization is converging, but it is unfortunate that sometimes it costs a lot of time to calculate it...

(Since I am not sure how well I explained my thinking, if my comment does not make any sense: I think I understand this code now; sorry for my confusion earlier, and let's not worry about my comment anymore. :))